pkg/asset/machines/azure/azuremachines.go

```diff
@@ -298,3 +329,20 @@ func CapzTagsFromUserTags(clusterID string, usertags map[string]string) (capz.Ta
 	}
 	return tags, nil
 }
+
+func bootDiagStorageURIBuilder(diag *aztypes.BootDiagnostics, cloudType aztypes.CloudEnvironment) string {
+	storageAccountURI := "https://%s.blob.%s"
+	if diag.Type == capz.UserManagedDiagnosticsStorage && diag.StorageAccountName != "" {
+		storageEndpointSuffix := azure.PublicCloud.StorageEndpointSuffix
+		switch cloudType {
+		case aztypes.USGovernmentCloud:
+			storageEndpointSuffix = azure.USGovernmentCloud.StorageEndpointSuffix
+		case aztypes.ChinaCloud:
+			storageEndpointSuffix = azure.ChinaCloud.StorageEndpointSuffix
+		case aztypes.GermanCloud:
+			storageEndpointSuffix = azure.GermanCloud.StorageEndpointSuffix
+		}
+		return fmt.Sprintf(storageAccountURI, diag.StorageAccountName, storageEndpointSuffix)
+	}
```

Comment:
Rather than import go-autorest (which is a deprecated package, but we still need overall as msft has no alternative for azure stack), you can get the storageEndpointSuffix from the `Environment` which is stored in the session:

```
session.Environment.StorageEndpointSuffix
```

`session` is not passed in to azuremachines, and perhaps that is best. You could add `storageEndpointSuffix` as an input value, although you will also have to rejigger `machines.go`